[quant][fx] Fix dynamic weighted op lowering when input is used multiple times

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #74365
* __->__ #74364
* #74231

Summary:
if a input is used multiple times in modules that are dynamically quantized:
```
x -- linear1
  \-- linear2
```
we'll insert quantize_per_tensor_dynamic and dequantize for input, and we'll have a duplicate pass
to duplicate dequantize ops for pattern matching:
```
x - quantize_per_tensor_dynamic - dequantize1 - linear1
                     \----- dequantize2 - linear2
```

But we also have a check in the lowering code that if quantize_per_tensor_dynamic is used by multiple nodes
we'll skip the pattern, so the pattern is not recognized, we need to duplicate quantize_per_tensor_dynamic as well in this case
to recover both patterns:
```
x - quantize_per_tensor_dynamic1 -- dequantize1 -- linear1
   \- quantize_per-tensor_dynamic2 -- dequantize2 -- linear2
```
so that they can be fused into dynamic linear:
```
x - linear_dynamic1
\-- linear_dynamic2
```

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_dynamic_linear_input_multiple_use

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34952755](https://our.internmc.facebook.com/intern/diff/D34952755)

jerryzh168 committed:
[quant][fx] Fix dynamic weighted op lowering when input is used multiple times

Summary:
if a input is used multiple times in modules that are dynamically quantized:
```
x -- linear1
  \-- linear2
```
we'll insert quantize_per_tensor_dynamic and dequantize for input, and we'll have a duplicate pass
to duplicate dequantize ops for pattern matching:
```
x - quantize_per_tensor_dynamic - dequantize1 - linear1
                     \----- dequantize2 - linear2
```

But we also have a check in the lowering code that if quantize_per_tensor_dynamic is used by multiple nodes
we'll skip the pattern, so the pattern is not recognized, we need to duplicate quantize_per_tensor_dynamic as well in this case
to recover both patterns:
```
x - quantize_per_tensor_dynamic1 -- dequantize1 -- linear1
   \- quantize_per-tensor_dynamic2 -- dequantize2 -- linear2
```
so that they can be fused into dynamic linear:
```
x - linear_dynamic1
\-- linear_dynamic2
```

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_dynamic_linear_input_multiple_use

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34952755).

jerryzh168 committed:
Update on "[quant][fx] Fix dynamic weighted op lowering when input is used multiple times"


Summary:
if a input is used multiple times in modules that are dynamically quantized:
```
x -- linear1
  \-- linear2
```
we'll insert quantize_per_tensor_dynamic and dequantize for input, and we'll have a duplicate pass
to duplicate dequantize ops for pattern matching:
```
x - quantize_per_tensor_dynamic - dequantize1 - linear1
                     \----- dequantize2 - linear2
```

But we also have a check in the lowering code that if quantize_per_tensor_dynamic is used by multiple nodes
we'll skip the pattern, so the pattern is not recognized, we need to duplicate quantize_per_tensor_dynamic as well in this case
to recover both patterns:
```
x - quantize_per_tensor_dynamic1 -- dequantize1 -- linear1
   \- quantize_per-tensor_dynamic2 -- dequantize2 -- linear2
```
so that they can be fused into dynamic linear:
```
x - linear_dynamic1
\-- linear_dynamic2
```

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_dynamic_linear_input_multiple_use

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34952755](https://our.internmc.facebook.com/intern/diff/D34952755)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx] Fix dynamic weighted op lowering when input is used multiple times"


Summary:
if a input is used multiple times in modules that are dynamically quantized:
```
x -- linear1
  \-- linear2
```
we'll insert quantize_per_tensor_dynamic and dequantize for input, and we'll have a duplicate pass
to duplicate dequantize ops for pattern matching:
```
x - quantize_per_tensor_dynamic - dequantize1 - linear1
                     \----- dequantize2 - linear2
```

But we also have a check in the lowering code that if quantize_per_tensor_dynamic is used by multiple nodes
we'll skip the pattern, so the pattern is not recognized, we need to duplicate quantize_per_tensor_dynamic as well in this case
to recover both patterns:
```
x - quantize_per_tensor_dynamic1 -- dequantize1 -- linear1
   \- quantize_per-tensor_dynamic2 -- dequantize2 -- linear2
```
so that they can be fused into dynamic linear:
```
x - linear_dynamic1
\-- linear_dynamic2
```

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_dynamic_linear_input_multiple_use

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34952755](https://our.internmc.facebook.com/intern/diff/D34952755)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34952755).

andrewor14 commented on torch/ao/quantization/fx/_convert_do_not_use.py:
```diff
@@ -125,6 +125,26 @@ def run_weight_observers(observed: GraphModule) -> None:
             # run the weight observer
             weight_observer_module()
 
+# this method is temporary will be removed soon
```
Is the real fix removing the check for whether dequantize nodes have multiple users in the lowering code? Is there a reason why we don't just directly change that instead of adding a temporary fix?

jerryzh168 committed:
Update on "[quant][fx] Fix dynamic weighted op lowering when input is used multiple times"


Summary:
if a input is used multiple times in modules that are dynamically quantized:
```
x -- linear1
  \-- linear2
```
we'll insert quantize_per_tensor_dynamic and dequantize for input, and we'll have a duplicate pass
to duplicate dequantize ops for pattern matching:
```
x - quantize_per_tensor_dynamic - dequantize1 - linear1
                     \----- dequantize2 - linear2
```

But we also have a check in the lowering code that if quantize_per_tensor_dynamic is used by multiple nodes
we'll skip the pattern, so the pattern is not recognized, we need to duplicate quantize_per_tensor_dynamic as well in this case
to recover both patterns:
```
x - quantize_per_tensor_dynamic1 -- dequantize1 -- linear1
   \- quantize_per-tensor_dynamic2 -- dequantize2 -- linear2
```
so that they can be fused into dynamic linear:
```
x - linear_dynamic1
\-- linear_dynamic2
```

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_dynamic_linear_input_multiple_use

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34952755](https://our.internmc.facebook.com/intern/diff/D34952755)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx] Fix dynamic weighted op lowering when input is used multiple times"


Summary:
if a input is used multiple times in modules that are dynamically quantized:
```
x -- linear1
  \-- linear2
```
we'll insert quantize_per_tensor_dynamic and dequantize for input, and we'll have a duplicate pass
to duplicate dequantize ops for pattern matching:
```
x - quantize_per_tensor_dynamic - dequantize1 - linear1
                     \----- dequantize2 - linear2
```

But we also have a check in the lowering code that if quantize_per_tensor_dynamic is used by multiple nodes
we'll skip the pattern, so the pattern is not recognized, we need to duplicate quantize_per_tensor_dynamic as well in this case
to recover both patterns:
```
x - quantize_per_tensor_dynamic1 -- dequantize1 -- linear1
   \- quantize_per-tensor_dynamic2 -- dequantize2 -- linear2
```
so that they can be fused into dynamic linear:
```
x - linear_dynamic1
\-- linear_dynamic2
```

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_dynamic_linear_input_multiple_use

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34952755](https://our.internmc.facebook.com/intern/diff/D34952755)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34952755).
